docs/proposals/scheduling/20220701-schedule-gang.md

```diff
@@ -0,0 +1,280 @@
+podA.Annotation[GangBundleName] = "ps"
+podA.Annotation[GangWaitTime] = "3600s"
+podA.Annotation[GangBundleMinRequiredNumber] = 5
+podA.Annotation[TotalChildrenNum] = 5
+podB.Annotation[GangName] = "job1-worker"
+podB.Annotation[GangBundleName] = "worker"
+podB.Annotation[GangWaitTime] = "3600s"
+podB.Annotation[GangBundleMinRequiredNumber] = 5
+podB.Annotation[TotalChildrenNum] = 5
+```
+### Implementation Details
+#### QueueSortPlugin
+We design a plugin to implement the QueueSort extension point separately, so that we can integrate the Queue logic of all our plugins and register them in k8s at one time.
+
+This time,we implement the Less function to arrange pods that belong to the same gang together.
+Specifically, on the basis of priority queuing，we set the following 3 rules:
+
+1.If both Pods are regularPods (ordinary Pods), the one  created firstly will be ranked first in the queue
+
+2.If one of the two Pods is a regularPod and the other is a gangPod (a Pod belonging to a Gang), we compare the creation time of the regularPod and the creation time of the Gang  which the gangPod belongs, the one created firstly will be ranked first in the queue.
+
+3.If both Pods are gangPods, we compare the creation times of the two gangs, the one created firstly will be ranked first in the queue. It is possible that the creation time of two Gangs is the same, and we compare their NamespacedName (the purpose here is to distinguish different Gangs).
+```go
+type QueueSortPlugin interface{
+    QueueSort(*QueuedPodInfo, *QueuedPodInfo) bool
+}
+```
+
+#### GangSchedulingPlugin
+##### target
+
+1.We should create a new gang-plugin to implement PreFilter\PostFilter\Permit\UnReserve stage. in this plugin, we should
+   provide gang-cache to store data and gang-controller to monitor gang status.
+
+2.We should implement gang-crd update\recover logic in gang-plugin
+
+##### detail
+We introduce Gang and BundleInfo for record gang status in scheduler memory. We especially explain ScheduleCycle and ChildrenScheduleRoundMap.
+These variables control bundle's scheduling cycle. at the beginning, ScheduleCycle is 1. when each pod comes to pre-filter, we will check pod's value in 
+ChildrenScheduleRoundMap is euqal to the  ScheduleCycle which means they are scheduled  at the same peace. If so, we continue to do the preFilter logic. If they are  not euqal, we set the pod's cycle in ChildrenScheduleRoundMap equal with ScheduleCycle. Finally,  when the last pod comes to 
+make all ChildrenScheduleRoundMap's value equal to ScheduleCycle, ScheduleCycle added by 1, which means a new schedule cycle.
+
+We continue to explain ScheduleCycleValid. in strict-mode, when a pod failed for scheduling, we will set ScheduleCycleValid to false in post-filter, which means in 
+this schedule cycle, the remaining pods should be rejected in pre-filter stage. when ScheduleCycle added by 1, we will reset the ScheduleCycleValid to true.
+```go
+type Gang struct {
```

Comment:
It is necessary to go in the same direction with upstream, because the open source project ecology must be compatible with the concept of the business community, so that the TF Job ecology is easier to connect to the Koordinator.

Some general suggestions:
1、Use the atomic concept of PodGroup
2、Make enhancements on top of PodGroup, such as adding multiple PodGroup gangScheduling semantics (similar to the Bundle concept here)
3、For users without complex requirements, no modifications from upstream to koordinator

Reply:
I think podgroup and gang are different,  podgroup means a batch of pods have same scheduling-policy.  gang's only limit the bind condition,  so in some features like `nvlink` or `gpu-pods-run-in-same-gpu-model-nodes`,  I think podgroup may be quoted.  Also as you mentioned,   if we introduce the bundle concept, the data-structure will be a little bigger different than PodGroup in community.

Reply:
`PodGroup` is a bad name. But it is indeed the API that the community currently uses to support Gang Scheduling. I also suggest we do enhancements on PodGroup.

Reply:
We alos should consider [lightweight-coscheduling](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/kep/2-lightweight-coscheduling) compatible with the community. We should consider the cooperation with the surrounding ecology to reduce the complexity of integrating Koordinator with the surrounding ecology.

Reply:
> podgroup means a batch of pods have same scheduling-policy

The PodGroup of the community is just a group and does not carry a scheduling policy.

Reply:
https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/kep/42-podgroup-coscheduling/README.md

Reply:
I modified as you wish.